horqrux/utils.py

```diff
@@ -151,12 +153,98 @@ def _jacobian(generator: Array, theta: float) -> Array:
 
 
 def _controlled(operator: Array, n_control: int) -> Array:
+    """
+    Create a controlled quantum operator with specified number of control qubits.
+
+    Args:
+        operator (jnp.ndarray): The base quantum operator to be controlled.
+        n_control (int): Number of control qubits.
+
+    Returns:
+        jnp.ndarray: The controlled quantum operator matrix
+    """
     n_qubits = int(log2(operator.shape[0]))
     control = jnp.eye(2 ** (n_control + n_qubits), dtype=default_dtype)
     control = control.at[-(2**n_qubits) :, -(2**n_qubits) :].set(operator)
     return control
 
 
+def controlled(
+    operator: jnp.ndarray,
+    target_qubits: TargetQubits,
+    control_qubits: ControlQubits,
+) -> jnp.ndarray:
+    """
+    Create a controlled quantum operator with specified control and target qubit indices.
+
+    Args:
+        operator (jnp.ndarray): The base quantum operator to be controlled.
+            Note the operator is defined only on `target_qubits`.
+        control_qubits (int or tuple of ints): Index or indices of control qubits
+        target_qubits (int or tuple of ints): Index or indices of target qubits
+
+    Returns:
+        jnp.ndarray: The controlled quantum operator matrix
+    """
+    controls: tuple = tuple()
+    targets: tuple = tuple()
+    if isinstance(control_qubits[0], tuple):
+        controls = control_qubits[0]
+    if isinstance(target_qubits[0], tuple):
+        targets = target_qubits[0]
+    nqop = int(log(operator.shape[0], 2))
+    ntargets = len(targets)
```

Comment:
```suggestion
    n_qop = int(log(operator.shape[0], 2))
    n_targets = len(targets)
```

I would suggest to add an underscore for every variable name that is supposed to mean "Number of ...". But this is merely stylistic.